### ISSUE TITLE
TypeError: list indices must be integers or slices, not DynamicJaxprTracer

### ISSUE THREAD

**Issue by Joy-Preetha** [google/jax#5839]
```python
@jit
def match_faces3(desc):
        distances = np.empty((len(desc), len(database)))
        time1 = time.time()
        for i, descr in enumerate(desc):
            for j, identity in enumerate(database):
                dist = []
                for k, id_desc in enumerate(identity[1]):
                    dist.append(jax_cosine(descr, id_desc))
                distances[i][j] = dist[jnp.argmin(jnp.asarray(dist))]
        time2 = time.time() - time1
        print("time2",time2)
        distances=distances.tolist()
        return jnp.asarray(distances)
```
<ipython-input-24-84d11a42c358> in match_faces3(desc)
     13                 for k, id_desc in enumerate(identity[1]):
     14                     dist.append(jax_cosine(descr, id_desc))
---> 15                 distances[i][j] = dist[jnp.argmin(jnp.asarray(dist))]
     16         time2 = time.time() - time1
     17         print("time2",time2)

TypeError: list indices must be integers or slices, not DynamicJaxprTracer

I understood using vectorization, it can be solved, but I dont know how we can implement vmap here.

**Comment by mattjj** [comment]
Thanks for raising this!

The issue is that the expression `dist[jnp.argmin(jnp.asarray(dist))]` is indexing the list referred to by `dist` with a JAX tracer. I wonder if we can make this error better by overloading `__index__`:

```python
In [1]: class A:
   ...:     def __index__(self):
   ...:         raise Exception("hi")
   ...:

In [2]: [1,2,3][A()]
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-2-543bfcd5aeff> in <module>
----> 1 [1,2,3][A()]

<ipython-input-1-cfb1eaf3bf37> in __index__(self)
      1 class A:
      2     def __index__(self):
----> 3         raise Exception("hi")
      4

Exception: hi
```

As for this particular code, there are actually a few issues with it we need to amend. One is this use of list indexing, but another is the in-place mutation of `distances`; JAX in general doesn't work with side-effects like in-place mutation of arrays (in this case a `numpy.ndarray`). The printing and timing are also side-effects which won't work in a jitted function.

I can provide some concrete advice if you extend it to be a runnable example (including example inputs).

**Comment by Joy-Preetha** [comment]
```python
database = [('8174',
  [array([ 9.2,  9.69,  7.8, -2.6,
           8.12, -5.39,  1.2,  2.2,
          -4.2, -2.21,  5.78,  1.9,
          -1.7,  1.05, -2.1, -1.6,
          -1.0, -3.1,  3.3, -4.6],
         dtype=float32)]),
 ('30819',
  [array([ 1.85, -7.70, -2.0,  5.65,
          -7.58,  3.3, -5.0, -7.25,
          -1.31,  8.06, -9.40, -1.2,
           1.4, -9.3,  7.8,  3.17,
          -1.2, -5.6,  3.39,  1.1],
         dtype=float32)]),
('41873',
   [array([ 1.7, -4.2, -7.5,  5.5,
          -5.6, -1.5,  5.2, -2.0,
          -1.13, -1.96, -1.7,  5.0,
           4.55, -8.0,  4.0, -9.86,
          -1.41, -5.4,  4.7, -1.28],
         dtype=float32)])]


descriptors = np.array([[-1.69,4.4,4.27,1.96,2.7,-5.73,-5.41,1.12,2.5,2.09,5.8,-8.7,6.7,4.1,7.4,6.3,9.7,2.4,6.4,3.3]])

@jit
def jax_cosine(u, v):
  return 1 - jnp.dot(u, v) / (jnp.linalg.norm(u) * jnp.linalg.norm(v))

```

**Comment by Joy-Preetha** [comment]
> ```python
> database = [('8174',
>   [array([ 9.2,  9.69,  7.8, -2.6,
>            8.12, -5.39,  1.2,  2.2,
>           -4.2, -2.21,  5.78,  1.9,
>           -1.7,  1.05, -2.1, -1.6,
>           -1.0, -3.1,  3.3, -4.6],
>          dtype=float32)]),
>  ('30819',
>   [array([ 1.85, -7.70, -2.0,  5.65,
>           -7.58,  3.3, -5.0, -7.25,
>           -1.31,  8.06, -9.40, -1.2,
>            1.4, -9.3,  7.8,  3.17,
>           -1.2, -5.6,  3.39,  1.1],
>          dtype=float32)]),
> ('41873',
>    [array([ 1.7, -4.2, -7.5,  5.5,
>           -5.6, -1.5,  5.2, -2.0,
>           -1.13, -1.96, -1.7,  5.0,
>            4.55, -8.0,  4.0, -9.86,
>           -1.41, -5.4,  4.7, -1.28],
>          dtype=float32)])]
> 
> 
> descriptors = np.array([[-1.69,4.4,4.27,1.96,2.7,-5.73,-5.41,1.12,2.5,2.09,5.8,-8.7,6.7,4.1,7.4,6.3,9.7,2.4,6.4,3.3]])
> 
> @jit
> def jax_cosine(u, v):
>   return 1 - jnp.dot(u, v) / (jnp.linalg.norm(u) * jnp.linalg.norm(v))
> ```

descriptors are desc

**Comment by jakevdp** [comment]
Does it make sense to continue this in the original discussion? https://github.com/google/jax/discussions/5821

**Comment by Joy-Preetha** [comment]
> Does it make sense to continue this in the original discussion? #5821

The discussion got closed and I couldnt reply in it anymore. I tried to understand vmap as you suggested but couldnt really find a way, so raised it as a new discussion since I couldnt continue in the previous one.

**Comment by jakevdp** [comment]
Sorry for the confusion  – it wasn't closed, it was converted from an issue to a discussion, because the latter format can work better for topics that are more about answering questions than reporting bugs. I think if you click on #5821, you'll see that the discussion remains open.

**Comment by Joy-Preetha** [comment]
Oh Thank you @jakevdp when I tried yesterday it was showing closed, but now I can see its open.

**Comment by jakevdp** [comment]
That was my mistake – I shouldn't have converted it without letting you know.

**Comment by mattjj** [comment]
I think Jake fixed this in #5853! Please reopen if I'm mistaken.
